// @sap/cds-dk/lib/init/constants.js
const os = require('os');

const MAVEN_ARCHETYPE_VERSION = '1.10.0';

const constants = {
    IS_WIN: (os.platform() === 'win32'),

    MAVEN_ARCHETYPE_VERSION,

    OPTION_HANA: 'hana',
    OPTION_JAVA: 'java',
    OPTION_MTA: 'mta',
    OPTION_NODEJS: 'nodejs',
    OPTION_CF_MANIFEST: 'cf-manifest',
    OPTION_PIPELINE: 'pipeline',
    OPTION_SAMPLES: 'samples',

    OPTION_JAVA_PACKAGE: 'java:package',
    OPTION_JAVA_MVN: 'java:mvn',

    // node enum
    PROJECT_TYPE: Object.freeze({ nodejs: 1, java: 2, unknown: 3 }),

    REGEX_JAVA_PACKAGE: /^[a-z_][a-z0-9_]*(\.[a-z_][a-z0-9_]*)*$/g,
    REGEX_PROJECT_NAME: /[^0-9a-zA-Z_-]/g,

    URLS: Object.freeze({
        PIPELINE_HELP: 'https://github.com/SAP/cloud-s4-sdk-pipeline',
        SAMPLES: 'https://github.com/SAP-samples/cloud-cap-samples',
        CAPIRE: 'https://cap.cloud.sap/docs/get-started',
        MAVEN_ARCHETYPE_HELP: `https://jarcasting.com/archetypes/com.sap.cds/cds-services-archetype/${MAVEN_ARCHETYPE_VERSION}/`,
        MAVEN_INSTALL_HELP: `https://maven.apache.org/install.html`
    }),

    PROJECT_FILES: ['package.json', 'pom.xml', '.cdsrc.json'],

    KIND_SQL: 'sql',
    KIND_HANA: 'hana'
}

constants.KINDS = [constants.KIND_HANA, constants.KIND_SQL];

module.exports = constants;
