// @sap/cds-dk/bin/cds.js
#!/usr/bin/env node

const cli = { //NOSONAR

  Shortcuts: {
    i: 'init',
    a: 'add',
    m: 'import',
    c: 'compile',
    s: 'serve',
    w: 'watch',
    e: 'env',
    b: 'build',
    d: 'deploy',
    l: 'login',
    o: 'logout',
    v: 'version', '-v':'version', '--version':'version',
    h: 'help', '?':'help', '-?':'help', '--help':'help',
    r: 'repl', '-r':'repl',
    '-e':'eval'
  },

  exec (cmd = process.argv[2], ...argv) {
    require('util').inspect.defaultOptions = { colors: !!process.stderr.isTTY, depth:11 }
    if (!argv.length) argv = process.argv.slice(3)
    if (!cmd) cmd = process.stdin.isTTY ? 'help' : 'compile'
    if (cmd in this.Shortcuts) cmd = process.argv[2] = this.Shortcuts[cmd]
    if (argv.some(a => this.Shortcuts[a]==='help'))  return this.help (cmd)

    const cds_cli = require (_local('@sap/cds/bin/cds'))  // this triggers loading of sap/cds lib
    DEBUG && DEBUG (`[cds] - @sap/cds ${global.cds.version} loaded: ${global.cds.home}`)
    if (cmd !== 'repl' && process.env.NODE_ENV !== 'test')  cds_cli.errorHandlers()

    const task = this.load (cmd, /*else*/ ()=> argv.unshift(cmd) && _defaultCompile (argv))
    if (task)  return task.apply (this, cds_cli.args(task,argv))
  },

  load (cmd,_default) {
    return /^(\.[/\\])?cds/i.test(cmd) && _dieForUnknownCommand (cmd)  // cds cds
    || _require ('./'+cmd)  // cds-dk commands
    || _require ('@sap/cds/bin/'+cmd, {paths:[process.cwd(), __dirname]}) // prefer local cds for the rest
    || _default && _default()
    || _dieForUnknownCommand (cmd)
  },

  help (cmd) { return this.exec ('help', cmd) }
}

const DEBUG = /\b(cli)\b/.test (process.env.DEBUG) && console.error

function _local (id) {
  return require.resolve (id, {paths:[process.cwd(), __dirname]})
}

function _defaultCompile (argv) {
  // Try all arguments if one resolves to a model.  If yes, we assume 'compile' is intended.
  const model = require (_local('@sap/cds')).resolve(argv)
  if (model)  return _require ('./compile')
  DEBUG && DEBUG (`[cds] - Command is not a model to compile: ${argv}`)
}

function _dieForUnknownCommand (cmd) {
  const commands = Array.from(new Set(Object.values(cli.Shortcuts))).sort().join(', ')
  console.error (`\n  Unknown command '${cmd}'\n\n  Available commands: ${commands}\n`)
  process.exit (1)
}

function _require (id,o) {
  try { var resolved = require.resolve(id,o) } catch(e){ DEBUG && DEBUG (`[cds] - Command not found: ${id}`); return }
  DEBUG && DEBUG (`[cds] - Command resolved: ${resolved}`)
  return require (resolved)
}

function _minVersion (id, minVersion) {
  const { version } = require (_local (id+'/package.json'))
  return version.split('.')[0] >= minVersion
}

if (_minVersion ('@sap/cds', 4)) {
  module.exports = Object.assign ((..._) => cli.exec(..._), cli)
}
else { // TODO compat to cds 3 remove after a while.
  module.exports = require (_local('@sap/cds/bin/cds'))
}
if (!module.parent)  module.exports()

/* eslint no-console:off */
