// @sap/cds-dk/bin/watch.js
module.exports = Object.assign ( watch, {
    flags: [ '--debug', '--in-memory', '--in-memory?' ],
    help: `
# SYNOPSIS

    *cds watch* [<project>]

    Tells cds to watch for relevant things to come or change in the specified
    project or the current work directory. Compiles and (re-)runs the server
    on every change detected, using the open-source package _nodemon_.

# OPTIONS

    - see _cds help serve_

# SEE ALSO

    Actually, *cds watch* is just a convenient shortcut for:
    *cds serve* --with-mocks --in-memory? --watch [--project <project>] ...
    Check out *cds serve ?* to learn more.

`})

const path = require ('path')
const t = module.exports.codes = {
    reset: '\x1b[0m', // Default
    bold: '\x1b[1m', // Bold/Bright
    link: '\x1b[4m', // underline
    red: '\x1b[91m', // Bright Foreground Red
    green: '\x1b[32m', // Foreground Green
    yellow: '\x1b[33m', // Foreground Green
    orange: '\x1b[38;2;255;140;0m' // darker orange, works with bright and dark background
}


function watch ([folder], {
    ext = 'cds,csn,csv,ts,mjs,cjs,js,json,properties,edmx,xml,env',
    args = ['serve', 'all', '--with-mocks'],
    ...argRest
}={}) {
    const inMemArg = Object.keys(argRest).find(a => /^in-memory\??$/.test(a))
    args.push(inMemArg ? `--${inMemArg}` : '--in-memory?')

    const log = (first,...more) => console.log (t.yellow + (first||''), ...more, t.reset)
    if (folder) {
        const cwd = process.env._original_cwd = process.cwd()
        try {
            process.chdir (folder)
            log (`cd ${folder}`)
        } catch(e){
            try {
                const resolved = path.dirname (require.resolve(folder+'/package.json', {paths: [process.cwd()]}))
                log (`cd ${path.relative(cwd,resolved)}`)
                process.chdir (resolved)
            }
            catch(_){ throw e }
        }
        process.on('exit', ()=> process.chdir (process.env._original_cwd))
    }
    log ()
    log (`${t.bold}cds ${args.join(' ')}`)
    log (`( watching: ${ext}... )`)
    let delayed = undefined
    const nodemon = require('../lib/unlike-nodemon')
    return nodemon ({
        watch: ['*','.env','.cdsrc.json'], ext, script:__filename, env: Object.assign ({
            NODE_PATH: path.resolve (__dirname, '../node_modules'), // allow global sqlite to be resolved
            _args: JSON.stringify(args),
            _original_cwd: process.env._original_cwd,
        }, process.env), options:argRest
    }).on('restart', (files)=>{
        clearTimeout (delayed)
        delayed = setTimeout(()=>{
            log (`${t.bold}        _______________________\n`)
            if (files) for (let each of files) {
                const [,ext] = /\.(\w+)$/.exec(each) || []
                for (let handle of FileHandlers [ext] || [])  handle (each)
            }
        }, 111)}
    ).on('quit', ()=>{
        log (`${t.bold+t.green}\n[cds] - my watch has ended.\n`)
        process.exit()
    })
}


const FileHandlers = {
    edmx: [ file => (FileHandlers._import || (FileHandlers._import = require ('./import'))) (file) ]
}

if (!module.parent) { // launched by nodemon
    process.env._cds_watch = true
    // run through cli in order to use its error handling
    return require('./cds') (...JSON.parse(process.env._args))
}
