// @sap/cds-dk/lib/init/template/java/index.js
const os = require('os');
const path = require('path');

const fse = require('@sap/cds-foss')('fs-extra')


const commandUtil = require('../../util/commandUtil');
const mvnArchetypeUtil = require('../../util/mvnArchetypeUtil');
const term = require('../../../util/term');

const TemplateBase = require('../templateBase');
const { URLS } = require('../../constants');


module.exports = class JavaTemplate extends TemplateBase {
    constructor(projectPath, generator) {
        super(projectPath, generator, __dirname);
    }

    getPriority() {
        return Number.MAX_VALUE;
    }

    async checkEnabled() {
        return true;
    }

    async run() {
        const mvnCmdArgs = mvnArchetypeUtil.getGenerateCmdArgs(this.projectName, this.options, this.logger);

        const tempFolder = await fse.mkdtemp(path.join(os.tmpdir(), `${this.projectName}_`));
        try {
            await commandUtil.spawnCommand('mvn', mvnCmdArgs.cmdLine, {
                cwd: tempFolder
            }, this.logger);

            await this.fsUtil.copy(path.join(tempFolder, mvnCmdArgs.artifactId), this.projectPath);
        } catch (err) {
            if (err.code === 'ENOENT' && err.path === 'mvn') {
                throw new Error(`Maven executable 'mvn' not found, follow ${term.info(URLS.MAVEN_INSTALL_HELP)} and install Maven on your machine.`);
            }
            throw err;
        } finally {
            await fse.remove(tempFolder);
        }
    }

    async finalize() {
        const relativeProjectPath = path.relative(this.cwd, this.projectPath);
        if (relativeProjectPath && relativeProjectPath !== '.') {
            this.logger.log(`Continue with 'cd ${relativeProjectPath}'`);
        }

        this.logger.log(`Learn about next steps at ${URLS.CAPIRE}`);
    }
}
